src/util/codec/mysql/decimal.rs

```diff
@@ -11,770 +11,2008 @@
+            }
+            if leading_digits > 0 {
+                src_word_start_idx -= 1;
+            }
+            src_int_word_cnt = int_word_cnt;
+            src_leading_digits = leading_digits;
+            res = Res::Overflow(());
+            error!("encode {} with prec {} and frac {} overflow",
+                   d.to_string(),
+                   prec,
+                   frac);
+        } else if int_size > src_int_size {
+            for _ in src_int_size..int_size {
+                write_u8!(self, mask as u8, written);
+            }
         }
-        truncated = &truncated[..pos];
 
-        if d.is_negative() {
-            try!(self.encode_i64(-exp as i64));
-            self.encode_bytes(truncated, true)
-        } else {
-            try!(self.encode_i64(exp as i64));
-            self.encode_bytes(truncated, false)
+        if frac_size < src_frac_size {
+            src_frac_word_cnt = frac_word_cnt;
+            src_trailing_digits = trailing_digits;
+            res = Res::Truncated(());
+            warn!("encode {} with prec {} and frac {} truncated",
+                  d.to_string(),
+                  prec,
+                  frac);
+        } else if frac_size > src_frac_size && src_trailing_digits > 0 {
+            if frac_word_cnt == src_frac_word_cnt {
+                src_trailing_digits = trailing_digits;
+                frac_size = src_frac_size;
+            } else {
+                src_frac_word_cnt += 1;
+                src_trailing_digits = 0;
+            }
+        }
+
+        if src_leading_digits > 0 {
+            let i = DIG_2_BYTES[src_leading_digits] as usize;
+            let x = (d.word_buf[src_word_start_idx] % TEN_POW[src_leading_digits]) ^ mask;
+            src_word_start_idx += 1;
+            write_word!(self, x, i, written);
+        }
+
+        let stop = src_word_start_idx + src_int_word_cnt as usize + src_frac_word_cnt as usize;
+        while src_word_start_idx < stop {
+            write_word!(self, d.word_buf[src_word_start_idx] ^ mask, 4, written);
+            src_word_start_idx += 1;
+        }
+
+        if src_trailing_digits > 0 {
+            let i = DIG_2_BYTES[src_trailing_digits];
+            let lim = if src_frac_word_cnt < frac_word_cnt {
+                DIGITS_PER_WORD as usize
+            } else {
+                trailing_digits
+            };
+            while src_trailing_digits < lim && DIG_2_BYTES[src_trailing_digits] == i {
+                src_trailing_digits += 1;
+            }
+            let x = (d.word_buf[src_word_start_idx] /
+                     TEN_POW[DIGITS_PER_WORD as usize - src_trailing_digits]) ^
+                    mask;
+            write_word!(self, x, i as usize, written);
+        }
+
+        if frac_size > src_frac_size {
+            for _ in (src_frac_size..frac_size).zip(written..(int_size + frac_size) as usize) {
+                write_u8!(self, mask as u8, written);
+            }
         }
+        Ok(res)
```

Comment:
Because rust using stream API to avoid creating another temporary buffer, so it can't change the first byte once written. Hence the logic is moved to macro `write_u8` and `write_word`.